File: 516. Longest Palindromic Subsequence-1.py
class Solution:
    def longestPalindromeSubseq(self, s: str) -> int:
        n = len(s)
        t = s[::-1]

        @cache
        def dfs(i, j):
            if i < 0:
                return 0
            if j < 0:
                return 0
            if s[i] == t[j]:
                return dfs(i - 1, j - 1) + 1
            else:
                return max(dfs(i - 1, j), dfs(i, j - 1))

        return dfs(n - 1, n - 1)

File: 516. Longest Palindromic Subsequence-1.cpp
class Solution {
public:
  vector<vector<int>> memo;
  int longestPalindromeSubseq(string s) {
    int n = s.size();
    string t = s;
    reverse(t.begin(), t.end());
    memo.assign(n, vector<int>(n, -1));

    return dfs(s, t, n - 1, n - 1);
  }

  int dfs(string &s, string &t, int i, int j) {
    if (i < 0 || j < 0)
      return 0;
    int &res = memo[i][j];
    if (res != -1)
      return res;
    if (s[i] == t[j])
      return res = dfs(s, t, i - 1, j - 1) + 1;
    else
      return res = max(dfs(s, t, i - 1, j), dfs(s, t, i, j - 1));
  }
};